File: src/algorithms/arrays/spiral-order.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> spiralOrder(vector<vector<int>> arr)
{
    if (arr.size() == 0)
        return {};

    vector<int> res;
    int row = 0, n = arr.size() - 1;
    int col = 0, m = arr[0].size() - 1;
    int dir = 0;

    while (row <= n && col <= m)
    {
        if (dir == 0)
        {
            for (int i = col; i <= m; i++)
                res.push_back(arr[row][i]);

            row++;
        }
        else if (dir == 1)
        {
            for (int i = row; i <= n; i++)
                res.push_back(arr[i][m]);

            m--;
        }
        else if (dir == 2)
        {
            for (int i = m; i >= col; i--)
                res.push_back(arr[n][i]);

            n--;
        }
        else if (dir == 3)
        {
            for (int i = n; i >= row; i--)
                res.push_back(arr[i][col]);

            col++;
        }

        dir = (dir + 1) % 4;
    }

    return res;
}

File: src-ts/algorithms/arrays/spiral-order.ts
function spiralOrder(arr: number[][]) {
  const res: number[] = [];

  let n = arr.length - 1;
  let m = arr[0].length - 1;
  let row = 0;
  let col = 0;
  let dir = 0;

  while (row <= n && col <= m) {
    switch (dir) {
      case 0: // l->r
        for (let c = col; c <= m; c++) res.push(arr[row][c]);
        row++;
        break;

      case 1: // u->d
        for (let r = row; r <= n; r++) res.push(arr[r][m]);
        m--;
        break;

      case 2: // r->l
        for (let c = m; c >= col; c--) res.push(arr[row][c]);
        n--;
        break;

      case 3: // d->u
        for (let r = n; r >= row; r--) res.push(arr[r][col]);
        col++;
        break;

      default:
        break;
    }

    dir = (dir + 1) % 4;
  }

  return res;
}
